src/reentrancy.sw

```diff
@@ -7,46 +7,56 @@ use ::context::contract_id;
 use ::auth::caller_is_external;
 use ::option::*;
 
+const SAVED_REGISTERS_OFFSET = 64; // 8 words * 8 bytes
+const CALL_FRAME_OFFSET = 48;      // 6 words * 8 bytes
+
 /// Returns `true` if the reentrancy pattern is detected, and `false` otherwise.
 pub fn is_reentrant() -> bool {
     let mut reentrancy = false;
     let mut internal = !caller_is_external();
 
-    // not sure about this yet
-    let mut call_frame_pointer = asm() {
-        fp: u64
-    }
+    let mut call_frame_pointer = get_current_frame_pointer();
 
     // Get our current contract ID
-    let current_id = contract_id();
+    let target_id = contract_id();
 
+    // reentrancy cannot happen in an external context. If we don't detect it by the time we get to an external context in the stack, then the reentrancy pattern is not present
     while internal {
-        let saved_registers_pointer = get_saved_regs_pointer(call_frame_pointer);
-        let temp_id = get_previous_contract_id(saved_registers_pointer)
-        if temp_id == current_id {
+        let previous_id = get_previous_contract_id(call_frame_pointer);
+
+        if previous_id == target_id {
             reentrancy = true;
             internal = false;
         } else {
             internal = !caller_is_external();
-            call_frame_pointer = saved_registers_pointer + 48;
+
+            call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
+
         };
     }
     reentrancy
 }
 
-const SAVED_REGISTERS_OFFSET = 64; // 8 words * 8 bytes
-const CALL_FRAME_OFFSET = 48;      // 6 words * 8 bytes
+// get a pointer to the current call frame
+fn get_current_frame_pointer() -> u64 {
+    asm() {
+        fp: u64
+    };
+}
```

Comment:
Could functions like this be moved to something like a `registers.sw`, that has functions to get every internal register's value?

Reply:
It could.
@otrho and I have discussed this a bit, not only to expose all the reserved registers, but even go as far as to add wrapper functions to all the common opcodes to cut down on the amount of `asm` we need to write/review/maintain as we build out libraries. 
Perhaps we should open a new ticket for that to discuss further.
For now, I will focus on making this correct (proved through testing) and then look to optimize/refactor as needed.